### PR TITLE
mgr: ceph fs status support json format

### PR DESCRIPTION
**Pull request by chenerqi**
ceph fs status -f json/json-prettry returns string format, add json/json-pretty format support.

Fixes: https://tracker.ceph.com/issues/42359
Signed-off-by: Erqi Chen <chenerqi@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
jenkins test submodules

File: src/pybind/mgr/status/module.py [hunk 1]
```diff
@@ -49,6 +51,13 @@ def get_rate(self, daemon_type, daemon_name, stat):
             return 0
 
     def handle_fs_status(self, cmd):
+        f_output = collections.OrderedDict()
```
Comment (tchaikov) [inline src/pybind/mgr/status/module.py]: why shall we order the output?

File: src/pybind/mgr/status/module.py [hunk 2]
```diff
@@ -49,6 +51,13 @@ def get_rate(self, daemon_type, daemon_name, stat):
...
+        f_type = None
```
Comment (tchaikov) [inline src/pybind/mgr/status/module.py]:
```suggestion
        format = cmd.get('format', 'plain')
```

File: src/pybind/mgr/status/module.py [hunk 3]
```diff
@@ -49,6 +51,13 @@ def get_rate(self, daemon_type, daemon_name, stat):
...
+        if "format" in cmd and (cmd['format'] == 'json' or 'json-pretty'):
```
Comment (tchaikov) [inline src/pybind/mgr/status/module.py]: i'd suggest just use `defaultdict`. or

```py
if format in ('json', 'json-pretty):
  json_output = {'mdsmap':[], 'pools':[], 'mdsversion':[]}
```

File: src/pybind/mgr/status/module.py [hunk 4]
```diff
@@ -109,17 +118,34 @@ def handle_fs_status(self, cmd):
...
-                        mgr_util.format_dimless(dns, 5),
-                        mgr_util.format_dimless(inos, 5)
-                    ])
-
+                    if f_type:
+                        mds = collections.OrderedDict()
+                        mds['rank'] = rank
+                        mds['name'] = info['name']
+                        mds['state'] = state
+                        mds['Reqs'] = mgr_util.format_dimless(
```
Comment (tchaikov) [inline src/pybind/mgr/status/module.py]: s/Reqs/rate/

File: src/pybind/mgr/status/module.py [hunk 5]
```diff
@@ -109,17 +118,34 @@ def handle_fs_status(self, cmd):
...
+                            self.get_rate("mds", info['name'], "mds_server.handle_client_request"),
+                            5, False) + "/s" if state == "active" else "0 /s"
```
Comment (tchaikov) [inline src/pybind/mgr/status/module.py]: remove the "/s" postfix.

File: src/pybind/mgr/status/module.py [hunk 6]
```diff
@@ -137,12 +161,26 @@ def handle_fs_status(self, cmd):
...
 
-                rank_table.add_row([
-                    "{0}-s".format(daemon_info['rank']), "standby-replay",
-                    daemon_info['name'], activity,
-                    mgr_util.format_dimless(dns, 5),
-                    mgr_util.format_dimless(inos, 5)
-                ])
+                if output_format in ('json', 'json-pretty'):
+                    json_output['mdsmap'].append(
+                        {
```
Comment (tchaikov) [inline src/pybind/mgr/status/module.py]: could you move `{` into the previous line just like `rank_table.add_row([`? the same applies to other places in your change.

File: src/pybind/mgr/status/module.py [hunk 7]
```diff
@@ -137,12 +161,26 @@ def handle_fs_status(self, cmd):
...
+                            'rank': rank,
+                            'name': daemon_info['name'],
+                            'state': 'standby-replay',
+                            'Evts': mgr_util.format_dimless(
+                                self.get_rate("mds", daemon_info['name'], "mds_log.replayed"),
+                                5, False),
+                            'dns': self.format_dimless(dns, 5, False),
+                            'inos': self.format_dimless(inos, 5, False)
+                        }
+                    )
```
Comment (tchaikov) [inline src/pybind/mgr/status/module.py]: and move `)` to the end of previous line.  the same applies to other places in your change.

File: src/pybind/mgr/status/module.py [hunk 8]
```diff
@@ -137,12 +157,24 @@ def handle_fs_status(self, cmd):
...
-                    daemon_info['name'], activity,
-                    mgr_util.format_dimless(dns, 5),
-                    mgr_util.format_dimless(inos, 5)
-                ])
+                if output_format in ('json', 'json-pretty'):
+                    json_output['mdsmap'].append({
+                        'rank': rank,
+                        'name': daemon_info['name'],
+                        'state': 'standby-replay',
+                        'Evts': mgr_util.format_dimless(
```
Comment (ideepika) [inline src/pybind/mgr/status/module.py]: s/Evts/events

File: src/pybind/mgr/status/module.py [hunk 9]
```diff
@@ -155,42 +187,80 @@ def handle_fs_status(self, cmd):
...
+                standby_table.add_row([standby['name']])
 
-        output += "\n" + standby_table.get_string() + "\n"
+        if output_format not in ('json', 'json-pretty'):
+            output += "\n" + standby_table.get_string() + "\n"
 
         if len(mds_versions) == 1:
-            output += "MDS version: {0}".format(list(mds_versions)[0])
+            if output_format in ('json', 'json-pretty'):
+                json_output['mdsversion'] = "{0}".format(list(mds_versions)[0])
```
Comment (ideepika) [inline src/pybind/mgr/status/module.py]: s/mdsversion/mds_version

File: src/pybind/mgr/status/module.py [hunk 10]
```diff
@@ -155,42 +187,80 @@ def handle_fs_status(self, cmd):
...
             for version, daemons in six.iteritems(mds_versions):
-                version_table.add_row([
-                    version,
-                    ", ".join(daemons)
-                ])
-            output += version_table.get_string() + "\n"
+                if output_format in ('json', 'json-pretty'):
+                    json_output['mdsversion'].append({
+                        'version': version,
+                        'daemons': ", ".join(daemons)
```
Comment (ideepika) [inline src/pybind/mgr/status/module.py]: you can use list to store multiple values. 
```
'daemons': [", ".join(daemons)]
```

File: src/pybind/mgr/status/module.py [hunk 11]
```diff
@@ -109,17 +112,34 @@ def handle_fs_status(self, cmd):
...
-                        mgr_util.format_dimless(dns, 5),
-                        mgr_util.format_dimless(inos, 5)
-                    ])
-
+                    if output_format in ('json', 'json-pretty'):
+                        json_output['mdsmap'].append({
+                            'rank': rank,
+                            'name': info['name'],
+                            'state': state,
+                            'Reqs': mgr_util.format_dimless(
```
[the remark this answers is not in the document]
Comment (chenerqi) [reply src/pybind/mgr/status/module.py]: 1. Updated.
2. Use rate and events to store formatted value, for both json and plain output.

**Comment by ideepika** [comment]
can you paste the output for the changes you have added?

**Comment by chenerqi** [comment]
> can you paste the output for the changes you have added?

# ceph fs status -f json-pretty
{
    "clients": [
        {
            "clients": 2,
            "fs": "cephfs"
        }
    ],
    "mdsmap": [
        {
            "dns": " 412k",
            "inos": " 412k",
            "name": "<hostname1>",
            "rank": 0,
            "rate": "   0 ",
            "state": "active"
        },
        {
            "name": "<hostname2>",
            "state": "standby"
        }
    ],
    "mds_version": "ceph version k2.0-internal(stable)",
    "pools": [
        {
            "avail": "32.8T",
            "id": 81,
            "name": "filesys_meta",
            "type": "metadata",
            "used": " 823M"
        },
        {
            "avail": "32.8T",
            "id": 82,
            "name": "filesys_data",
            "type": "data",
            "used": "1445M"
        }
    ]
}

**Comment by ideepika** [comment]
> > can you paste the output for the changes you have added?
> 
> ```
> # ceph fs status -f json-pretty
> {
>     "clients": [
>         {
>             "clients": 2,
>             "fs": "cephfs"
>         }
>     ],
>     "mdsmap": [
>         {
>             "dns": " 412k",
>             "inos": " 412k",
>             "name": "<hostname1>",
>             "rank": 0,
>             "rate": "   0 ",
>             "state": "active"
>         },
>         {
>             "name": "<hostname2>",
>             "state": "standby"
>         }
>     ],
>     "mds_version": "ceph version k2.0-internal(stable)",
>     "pools": [
>         {
>             "avail": "32.8T",
>             "id": 81,
>             "name": "filesys_meta",
>             "type": "metadata",
>             "used": " 823M"
>         },
>         {
>             "avail": "32.8T",
>             "id": 82,
>             "name": "filesys_data",
>             "type": "data",
>             "used": "1445M"
>         }
>     ]
> }
> ```

1. there's extra spaces in output for `"dns": "  10 ", `"inos": "  13 "` `"rate": "   0 ",` `"used": " 384k"` can you make update all instances of such values.
2. think the use of JSON format should be more relevant for machines in which case, instances like 
1011G or 384k could output a completely numeric value like 384000 

thanksd for your contribution, can you make those updates please! :)

**Comment by chenerqi** [comment]
> > > can you paste the output for the changes you have added?
> > 
> > 
> > ```
> > # ceph fs status -f json-pretty
> > {
> >     "clients": [
> >         {
> >             "clients": 2,
> >             "fs": "cephfs"
> >         }
> >     ],
> >     "mdsmap": [
> >         {
> >             "dns": " 412k",
> >             "inos": " 412k",
> >             "name": "<hostname1>",
> >             "rank": 0,
> >             "rate": "   0 ",
> >             "state": "active"
> >         },
> >         {
> >             "name": "<hostname2>",
> >             "state": "standby"
> >         }
> >     ],
> >     "mds_version": "ceph version k2.0-internal(stable)",
> >     "pools": [
> >         {
> >             "avail": "32.8T",
> >             "id": 81,
> >             "name": "filesys_meta",
> >             "type": "metadata",
> >             "used": " 823M"
> >         },
> >         {
> >             "avail": "32.8T",
> >             "id": 82,
> >             "name": "filesys_data",
> >             "type": "data",
> >             "used": "1445M"
> >         }
> >     ]
> > }
> > ```
> 
> 1. there's extra spaces in output for `"dns": "  10 ", "inos": "  13 " "rate": "   0 ", "used": " 384k"` can you make update all instances of such values.
> 2. think the use of JSON format should be more relevant for machines in which case, instances like
>    1011G or 384k could output a completely numeric value like 384000
> 
> thanks for your contribution, can you make those updates please!

updated and latest output as follow.
```
# sudo ceph fs status -f json-pretty
{
    "clients": [
        {
            "clients": 2,
            "fs": "cephfs"
        }
    ],
    "mds_version": "ceph version k2.0-internal(stable)",
    "mdsmap": [
        {
            "dns": 412135,
            "inos": 412088,
            "name": "<hostname1>",
            "rank": 0,
            "rate": 0.0,
            "state": "active"
        },
        {
            "name": "<hostname2>",
            "state": "standby"
        }
    ],
    "pools": [
        {
            "avail": 36122218463232,
            "id": 81,
            "name": "filesys_meta",
            "type": "metadata",
            "used": 864465786
        },
        {
            "avail": 36122218463232,
            "id": 82,
            "name": "filesys_data",
            "type": "data",
            "used": 1516037183
        }
    ]
}
```

**Comment by ideepika** [comment]
jenkins test make check arm64

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2020-01-14_13:07:57-rados-wip-kefu-testing-2020-01-14-1719-distro-basic-smithi/

**Pull request merged**